File: backend/app/schemas/trip.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field, computed_field, model_validator

from app.models.enums import TripStatus
from app.schemas.assignment import AssignmentDriver
from app.schemas.driver import AssignedVehicle

Latitude = Field(default=None, ge=-90, le=90)
Longitude = Field(default=None, ge=-180, le=180)


class TripCreate(BaseModel):
    vehicle_id: UUID
    driver_id: UUID
    source: str = Field(min_length=2, max_length=200)
    destination: str = Field(min_length=2, max_length=200)
    scheduled_start: datetime
    scheduled_end: datetime
    notes: str | None = None

    @model_validator(mode="after")
    def _ordered(self) -> "TripCreate":
        if self.scheduled_end <= self.scheduled_start:
            raise ValueError("scheduled_end must be after scheduled_start")
        return self


class TripStart(BaseModel):
    start_odometer: int = Field(ge=0)
    latitude: Decimal | None = Latitude
    longitude: Decimal | None = Longitude


class TripComplete(BaseModel):
    end_odometer: int = Field(ge=0)
    latitude: Decimal | None = Latitude
    longitude: Decimal | None = Longitude
    notes: str | None = None


class TripStatusUpdate(BaseModel):
    status: TripStatus


class TripResponse(BaseModel):
    id: UUID
    trip_number: str
    vehicle: AssignedVehicle
    driver: AssignmentDriver
    source: str
    destination: str
    scheduled_start: datetime
    scheduled_end: datetime
    status: TripStatus
    actual_start: datetime | None
    actual_end: datetime | None
    start_odometer: int | None
    end_odometer: int | None
    start_latitude: Decimal | None
    start_longitude: Decimal | None
    end_latitude: Decimal | None
    end_longitude: Decimal | None
    distance_km: Decimal | None
    notes: str | None
    created_at: datetime

    @computed_field
    @property
    def duration_minutes(self) -> int | None:
        if self.actual_start and self.actual_end:
            return int((self.actual_end - self.actual_start).total_seconds() // 60)
        return None

File: ios/Fleet Management Test/Models/Trip.swift
//
//  Trip.swift
//  Fleet Management Test
//

import Foundation

/// A trip. Almost everything describing what *actually* happened is null until
/// the trip runs, so a `SCHEDULED` trip arrives with ~9 null fields.
struct Trip: Decodable, Sendable, Identifiable, Hashable {
    let id: String
    let tripNumber: String
    let vehicle: TripVehicle?
    let driver: TripDriver?
    let source: String
    let destination: String
    let scheduledStart: Date
    let scheduledEnd: Date
    let status: TripStatus
    let actualStart: Date?
    let actualEnd: Date?
    let startOdometer: Int?
    let endOdometer: Int?

    // Decimals arrive as strings so they never round through a float.
    // Kept as strings for display; parsed only where arithmetic is needed.
    let startLatitude: String?
    let startLongitude: String?
    let endLatitude: String?
    let endLongitude: String?
    let distanceKm: String?

    let notes: String?
    let createdAt: Date
    let durationMinutes: Int?

    var route: String { "\(source) → \(destination)" }

    var startCoordinate: Coordinate? {
        Coordinate(latitude: startLatitude, longitude: startLongitude)
    }

    var endCoordinate: Coordinate? {
        Coordinate(latitude: endLatitude, longitude: endLongitude)
    }
}

/// The trimmed vehicle the trips endpoints embed - not a full `Vehicle`.
struct TripVehicle: Decodable, Sendable, Hashable {
    let id: String
    let registrationNumber: String
    let make: String
    let model: String

    var makeModel: String { "\(make) \(model)" }
}

struct TripDriver: Decodable, Sendable, Hashable {
    let id: String
    let fullName: String
    let licenseNumber: String
}

enum TripStatus: String, APIEnum {
    case scheduled = "SCHEDULED"
    case started = "STARTED"
    case inProgress = "IN_PROGRESS"
    case completed = "COMPLETED"
    case cancelled = "CANCELLED"
    case unknown = "UNKNOWN"

    var displayName: String {
        switch self {
        case .scheduled: return "Scheduled"
        case .started: return "Started"
        case .inProgress: return "In progress"
        case .completed: return "Completed"
        case .cancelled: return "Cancelled"
        case .unknown: return "Unknown"
        }
    }
}

/// One GPS ping. `GET /trips/{id}/route` returns these as a plain array,
/// oldest first - no pagination envelope.
struct RoutePoint: Decodable, Sendable, Identifiable, Hashable {
    let id: String
    let tripId: String
    let vehicleId: String
    let latitude: String
    let longitude: String
    let speedKmph: String?
    let heading: String?
    let accuracyM: String?
    let recordedAt: Date
    let receivedAt: Date

    var coordinate: Coordinate? { Coordinate(latitude: latitude, longitude: longitude) }
}
